designcompose/src/main/java/com/android/designcompose/squoosh/SquooshAnimate.kt

```diff
@@ -132,173 +145,240 @@ internal class SquooshAnimationControl(
+    val transition: Transition,
+    val action: AnimatedAction?,
+    val variant: VariantAnimationInfo?,
+    var animationControl: SquooshAnimationControl? = null
+)
+
+/// Create a new tree, based on "from", which additional nodes from "to" where an animation has
+/// been requested. If `requestedAnimationControls` is empty or has no matching ids then the output
+/// tree is the same as `from`.
+///
+/// The tree that's returned from this function should only be used for rendering and event
+/// dispatching. Other algorithms like layout won't work on this tree, because it's combining
+/// elements of two other trees.
+internal fun createMergedAnimationTree(
     from: SquooshResolvedNode,
-    fromNodeId: String,
     to: SquooshResolvedNode,
-    toNodeId: String
-): SquooshAnimationControl? {
-    val items = ArrayList<SquooshAnimatedItem>()
-    val start = findNode(from, fromNodeId)
-    val end = findNode(to, toNodeId)
-    if (start != null && end != null) {
-        mergeTreesWithAnimation(start, end, items)
-        return SquooshAnimationControl(root = to, items = items)
+    requestedAnimations: HashMap<String, SquooshAnimationRequest>,
+    parent: SquooshResolvedNode? = null,
+): SquooshResolvedNode {
+    if (requestedAnimations.isEmpty()) return from.cloneSelfAndChildren(parent)
+
+    // Does this match a thing that we need to animate?
+    var requestedAnim = requestedAnimations[from.view.id]
+    if (requestedAnim == null) requestedAnim = requestedAnimations[from.unresolvedNodeId]
+    val (cloned: SquooshResolvedNode, alreadyBuiltChildren) =
+        if (requestedAnim != null && requestedAnim.animationControl == null) {
+            // Can we find it in the "to" tree?
+            val toNode = findNode(to, requestedAnim.toNodeId)
+            if (toNode != null) {
+                val animations: ArrayList<SquooshAnimatedItem> = ArrayList()
+                val mergedClonedNode = mergeRecursive(from, toNode, parent, animations)
+                if (animations.isNotEmpty())
+                    requestedAnim.animationControl =
+                        SquooshAnimationControl(mergedClonedNode, animations)
+
+                Pair(mergedClonedNode, true)
+            } else {
+                // An animation was requested, but no destination node could be found. We'll end up
+                // leaving the animationControl field blank, which will result in an error log later
+                // on.
+                Pair(from.cloneSelf(parent), false)
+            }
+        } else {
+            // Clone the target node, then look at its siblings, then at its child.
+            Pair(from.cloneSelf(parent), false)
+        }
+
+    // First recurse along siblings
+    val nextFrom = from.nextSibling
+    if (nextFrom != null) {
+        val nextCloned = createMergedAnimationTree(nextFrom, to, requestedAnimations, parent)
+        // Insert at the end; does this reorder? Hope not
+        var c: SquooshResolvedNode? = cloned
```

Comment:
Odd that this needs `?`

Reply:
It's because I do `c = c.nextSibling` even though it's guarded. I think this linked list code is really quite hard going, although it seems to run fairly fast. Maybe if we move it to Rust we can use the same thing `expanse` uses for its tree and it'll be fast enough without being so hard to read?